File: LeetCode/Java Solutions/Design a Food Rating System/design.java
import java.util.*;

public class FoodRatings {
    private Map<String, String> foodToCuisine;
    private Map<String, Integer> foodToRating;
    private Map<String, TreeMap<Integer, Set<String>>> cuisineToFood;

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        foodToCuisine = new HashMap<>();
        foodToRating = new HashMap<>();
        cuisineToFood = new HashMap<>();

        for (int i = 0; i < foods.length; i++) {
            foodToCuisine.put(foods[i], cuisines[i]);
            foodToRating.put(foods[i], ratings[i]);
            cuisineToFood.putIfAbsent(cuisines[i], new TreeMap<>(Collections.reverseOrder()));
            cuisineToFood.get(cuisines[i]).putIfAbsent(ratings[i], new TreeSet<>());
            cuisineToFood.get(cuisines[i]).get(ratings[i]).add(foods[i]);
        }
    }

    public void changeRating(String food, int newRating) {
        String cuisine = foodToCuisine.get(food);
        int oldRating = foodToRating.get(food);
        foodToRating.put(food, newRating);

        cuisineToFood.get(cuisine).get(oldRating).remove(food);
        if (cuisineToFood.get(cuisine).get(oldRating).isEmpty()) {
            cuisineToFood.get(cuisine).remove(oldRating);
        }

        cuisineToFood.putIfAbsent(cuisine, new TreeMap<>(Collections.reverseOrder()));
        cuisineToFood.get(cuisine).putIfAbsent(newRating, new TreeSet<>());
        cuisineToFood.get(cuisine).get(newRating).add(food);
    }

    public String highestRated(String cuisine) {
        Set<String> foods = cuisineToFood.get(cuisine).firstEntry().getValue();
        return foods.iterator().next();
    }
}

File: LeetCode/Python Solutions/Design a Food Rating System/design.py
from typing import List
from collections import defaultdict
import heapq

class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.food_info = {food: [cuisine, rating] for food, cuisine, rating in zip(foods, cuisines, ratings)}
        self.cuisine_max_heap = defaultdict(list)
        for food, (cuisine, rating) in self.food_info.items():
            heapq.heappush(self.cuisine_max_heap[cuisine], (-rating, food))

    def changeRating(self, food: str, newRating: int) -> None:
        cuisine, oldRating = self.food_info[food]
        self.food_info[food][1] = newRating
        heapq.heappush(self.cuisine_max_heap[cuisine], (-newRating, food))

    def highestRated(self, cuisine: str) -> str:
        while self.cuisine_max_heap[cuisine]:
            rating, food = self.cuisine_max_heap[cuisine][0]
            if self.food_info[food][1] == -rating:
                return food
            heapq.heappop(self.cuisine_max_heap[cuisine])
